Remove the need to attempt `SchemaConfigModeAttr` fixups on all blocks

jbardin opened the pull request:
In order to handle the possibility of providers using `SchemaConfigModeAttr`, Terraform currently must attempt to treat all configuration blocks as possible attributes. This was relatively safe up until now, because the legacy schema could not define complex attribute types which might match any block structures.

Now that we have an SDK capable of using the full Terraform type system, resources can easily create attributes which could match the structure of natural HCL blocks. Rather than risk provider developers discovering this fixup behavior and relying on it to write certain styles of configuration, we can attempt to prevent the fixup altogether when possible. We can also remove the overhead of running the fixup on every decode operation (workaround added for #25889), and the possibility of it further conflicting with other features like `dynamic` blocks (as seen in #22340).

Because the new SDK is designed with the intent of muxing legacy provider resources along with new-style resources to aid in transitioning providers to the new SDK, we cannot solely rely on the protocol version to indicate if the fixup needs to be supported. Instead we add a `Capabilities` field to the resource schema, so that the SDK can opt-in to the more strict behavior. The `Capabilities` field can be extended with other options if needed during the protocol 6 lifecycle, but will initially only contain the single `StructuralAttrs` option.

apparentlymart commented on docs/plugin-protocol/tfplugin6.1.proto:
```diff
@@ -0,0 +1,332 @@
+    // Current valid values are
+    // 0: Legacy protocol5 compatible resource
+    // 1: Protocol6 resource which will not use SchemaConfigModeAttr
+    int64 capability = 3;
```
I like the idea of capability negotiation, but it feels weird to pack it all into just a single 64-bit integer like this, cause that puts constraints on how we could use it in future.

What do you think about defining a `message Capability` which for now just has a single field `bool structural_attrs` which, when enabled, turns on the ability to use structural attribute types in return for _disabling_ the legacy attributes as blocks mode?

(I'm implicitly suggesting here that it would be invalid to use `Attribute.nested_type` _without_ activating this capability, so that these two behaviors are forced to be mutually exclusive and thus we can avoid dealing with the inevitable quirky interactions between these two modes -- but I feel more strongly about using a message type rather than and integer than I do about what exactly the flag inside does.)

Then if we have new capabilities in future we can add them as separate fields with their own docs. Some of them might have types other than `bool`, depending on what we end up needing to represent.

jbardin replied:
I guess we need to decide if we actually want ever add more capabilities based via the schema alone. I guess I brought this on by picking the work `Capability` over some form of "MetaVersion" :D  If we do actually want "capabilities" stored in here in the long run, then I see no reason to not add this is an a new message type. 

I don't think we can force this to be opt-in for the _new_ behavior however, since there could technically be providers built using the existing mechanisms which would be broken by disabling nested types, and the default zero value should not exclude those. Though if you think we can get away with that somehow, I am all for making the behavior as strict as possible.

jbardin replied:
Since this is only prevent undesirable but not necessarily broken behavior, I see no need to backport this to 1.0, which gives the SDKs time to implement and get releases in place to opt-in. It does feel kind of strange that we implemented a new protocol to add the additional features, only to require an additional flag to opt-in to said features, but it is really to exclude old behavior to it's not _that_ awful.

If that solution is ok with @paddycarver, I'll add `Schema.Capabilities.NestedTypes` to the protocol and re-work the PR accordingly.

apparentlymart replied:
My sense was that the protocol version change handled the other end of this negotiation, preventing odd behavior if someone tries to use a newer provider with an older version of Terraform. In retrospect it seems like we could've potentially added a more fine-grain mechanism for _that_ too, but the new protocol version gave us an opportunity to clean up some other oddities we learned about since 0.12 anyway so I think it was overall worth the effort.

jbardin replied:
And here I missed the less vague term "StructuralAttrs" :/, the PR cleanup will also `s/NestedTypes/StructuralAttrs/`

jbardin commented:
Updated the PR so we can see what the new idea looks like. There's probably a bit of cleanup to do in the PR after the refactor, but I can attend to that after we agree on the new structure.

Changed the protocol to add `Schema.Capabilities` for future extensions. Added the `Capabilities.NestedType` field which will enables nested attribute object types, while implicitly disabling `SchemaConfigModeAttr`.

apparentlymart reviewed (approved):
I'm going to leave the final call on what we do about the window of existing provider releases that might've used the new behavior without opting into it to the SDK team, so this is a partial/conditional approval in the sense that what you've got here seems technically reasonable and gives us an extensible way to do feature negotiation in future.

If we end up tweaking the details here in a way that's still in the spirit of having a capability negotiation message (if the details of what it's negotiating end up different) then I'm happy for this to move forward without me reviewing it another time.

paddycarver commented:
Hey y'all, thanks for looping me in on this.

Is this something that needs addressed immediately, or is "by the end of next week" an okay timeline? I'd like to loop the rest of @hashicorp/terraform-enablement-sdk in on this to discuss it as a team, but if this is time-sensitive we can figure something out.

I specifically want to dig into the implications of this a bit more:

> (I'm implicitly suggesting here that it would be invalid to use Attribute.nested_type without activating this capability, so that these two behaviors are forced to be mutually exclusive and thus we can avoid dealing with the inevitable quirky interactions between these two modes -- but I feel more strongly about using a message type rather than and integer than I do about what exactly the flag inside does.)

jbardin commented:
I think a week or so is fine, since we are planning the change for v1.1.

kmoe commented on docs/plugin-protocol/tfplugin6.1.proto:
```diff
@@ -0,0 +1,334 @@
+// Terraform Plugin RPC protocol version 6.0
```
```suggestion
// Terraform Plugin RPC protocol version 6.1
```

further non-typo feedback will be given through another medium ;)

kmoe commented on docs/plugin-protocol/tfplugin6.1.proto:
```diff
@@ -0,0 +1,334 @@
+// Terraform Plugin RPC protocol version 6.0
+//
+// This file defines version 6.0 of the RPC protocol. To implement a plugin
```
```suggestion
// This file defines version 6.1 of the RPC protocol. To implement a plugin
```

paddycarver reviewed (approved):
Minus the comments from @kmoe above, looks good to me. 👍

jbardin commented:
Due to the uncertainty and risk around changing the protocol requirements at this stage, I'm going to close this PR.
A heuristic-based approach like #29624 may turn out to be sufficient for out needs, else we will need an optional way for the provider to opt-in to signaling which resources were generated by the legacy SDK.
